File: KE_CPP/Minigrep/Minigrep.cpp
#include "ExpandableBitVec.h"
#include <fstream>
#include <string>
#include <regex>
#include <stdlib.h>

int main(int argc, char* argv[]) {
	switch (argc) {
	case 1:
		fprintf(stderr, "You need to pass in a file as the first argument.");
		return -1;
	case 2:
		fprintf(stderr, "You need to pass in a pattern to look for as the second argument.");
		return -1;
	}

	std::regex pat(argv[2]);
	ExpandableBitVec bv(256);

	std::ifstream file("jabberwocky.txt");
	std::string line;
	while (std::getline(file, line)) {
		auto match = std::smatch{};
		if (std::regex_search(line, match, pat)) {
			auto prefix = match.prefix();

			for (auto it = match.begin(); it != match.end(); ++it) {
				int start = (*it).length();
				int end = start + (*it).str().length();
				printf("Going from %d to %d\n", start, end);
				for (int i = start; i < end; i++) {
					bv.Set(i);
				}
			}
		}

		for (int col = 0; col < line.length(); col++) {
			if (bv.Index(col)) {
				printf("\x1B[32;1m%c\x1B[37m", line.at(col));
			}
			else {
				printf("%c", line.at(col));
			}
		}
		printf("\n");
		bv.Clear();
	}
}

File: KE_Python/Minigrep/minigrep.py
from expandable_bit_vec import ExpandableBitVec
import sys
import re

if __name__ == "__main__":
    argv = sys.argv
    argc = len(argv)

    if argc == 1:
        print("You need to pass in a file as the first argument.")
        sys.exit(-1)
    elif argc == 2:
        print("You need to pass in a pattern to look for as the second argument.")
        sys.exit(-1)

    pattern = re.compile(argv[2])

    toBePrinted = ""
    with open(argv[1], 'r', encoding='utf-8') as file:
        ebv = ExpandableBitVec(256)
        for line in file.readlines():
            for match in pattern.finditer(line):
                for i in range(match.start(), match.end()):
                    ebv[i] = True

            for i, c in enumerate(line):
                if ebv[i]:
                    toBePrinted += f"\033[32m{line[i]}\033[0m"
                else:
                    toBePrinted += f"{line[i]}"
            ebv.clear()

    print(toBePrinted)